scripts/generate_rust_analyzer.py

```diff
@@ -0,0 +1,175 @@
+    ret = []
+    with open( objtree + "/include/generated/rustc_cfg") as f:
+        for line in f:
+            line = line.replace('--cfg=', '')
+            line = line.replace('\n', '')
+            ret.append(line)
+    return ret
+
+def get_default_deps(source = crate_list):
+    return generate_deps(source, ["core", "alloc", "kernel"])
+
+def generate_crate(name, module, member = True, deps = get_default_deps(), cfg = generate_cfgs(), extra = {}, edition = "2018"):
+    return {
+        "display_name": name,
+        "root_module": module,
+        "edition": edition,
+        "is_workspace_member": member,
+        "cfg": cfg,
+        "deps": deps,
+        "env": {
+            "RUST_MODFILE": "This is only for rust-analyzer"
+        }
+    }
+
+def generate_kernel(srctree = srctree, objtree = objtree, lib_src = lib_src, bindgen_file = bindgen_file):
+    ret = []
+
+    # core
+    ret.append(
+        generate_crate(
+            name="core",
+            module=lib_src + "/core/src/lib.rs",
+            member=False,
+            deps=[],
+            cfg=[]
+        )
+    )
+
+    # compiler_builtins
+    ret.append(
+        generate_crate(
+            name="compiler_builtins",
+            module=srctree + "/rust/compiler_builtins.rs",
+            deps=[],
+            cfg=[]
+        )
+    )
+
+    # alloc
+    ret.append(
+        generate_crate(
+            name="alloc",
+            module=lib_src + "/alloc/src/lib.rs",
+            member=False,
+            deps=generate_deps(ret, ["core", "compiler_builtins"]),
+            cfg=[]
+        )
+    )
+
+    # module
+    module = generate_crate(
+        name="module",
+        module=srctree + "/rust/module.rs",
+        deps=[],
+        cfg=[],
+    )
+    module["proc_macro_dylib_path"] = "rust/libmodule.so"
+    ret.append(
+        module
+    )
+
+    # kernel
+    kernel = generate_crate(
+        name="kernel",
+        module=srctree + "/rust/kernel/lib.rs",
+        deps=generate_deps(ret, ["core", "alloc", "module"]),
+    )
+    kernel["env"]["RUST_BINDINGS_FILE"] = bindgen_file
+    kernel["source"] = {
+        "include_dirs": [ srctree + "/rust/kernel", objtree + "/rust" ],
+        "exclude_dirs": []
+    }
+    ret.append(kernel)
+
+    return ret
+
+def does_file_contains(filepath, search):
+    with open (filepath) as f:
+        for line in f:
+            if search in line:
+                return True
+    return False
+
+def check_and_generate(filepath, source=crate_list):
+    file = path.basename(filepath)
+    makefile = path.dirname(filepath) + "/Makefile"
+    name = file.replace('.rs', '')
+    objname = name + ".o"
+    print("checking " + objname, file=sys.stderr)
+    if does_file_contains(makefile, objname):
+        print("building crate " + name, file=sys.stderr)
+        return generate_crate(
+            name=name,
+            module=filepath,
+            deps=get_default_deps(source=source)
+        )
+
+
+def find_rs(tree):
```

Comment:
A feature you may like for `find_rs` is generators, i.e. using `yield` to return values one-by-one instead of returning a full list, similar to iterators in other languages.

Reply:
Not sure how I would implement this